modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -2011,13 +2011,135 @@ def window(
         PandasDataframe
             A new PandasDataframe with the reduce function applied over windows of the specified
                 axis.
-
-        Notes
-        -----
-        The user-defined reduce function must reduce each window’s column
-        (row if axis=1) down to a single value.
         """
-        pass
+
+        axis = Axis(axis)
+
+        # applies reduction function over entire virtual partition
+        def window_function_complete(virtual_partition):
+            virtual_partition_copy = virtual_partition.copy()
+            window_result = reduce_fn(virtual_partition_copy)
+            return window_result
+
+        # applies reduction function over one window of virtual partition
+        def window_function_partition(virtual_partition):
+            virtual_partition_copy = virtual_partition.copy()
```

Comment:
```suggestion
	    # have to copy the pandas dataframe on ray because it's immutable
            virtual_partition_copy = virtual_partition.copy()
```